ruby/bindings/InitRubyBindings.cpp

```diff
@@ -920,6 +945,37 @@ void locateEmbeddedGems(bool use_bundler) {
 
     $logger.info "Bundling with groups [#{keep_groups.join(',')}]"
 
+    # Filter out dependencies with native extensions that we already have embedded in the CLI and that satisfy the requirements
+    locked_specs = Bundler.definition.instance_variable_get(:@locked_specs)
+
+    embedded_gems_to_activate.each do |spec|
+      next if spec.extensions.empty?
+
+      next unless locked_specs.any? {|locked_spec| locked_spec.name == spec.name}
+
+      dep_to_requirements = {}
+      locked_specs.each do |locked_spec|
+        deleted_deps = locked_spec.dependencies.select { |dep| dep.name == spec.name && dep.requirement.satisfied_by?(spec.version) }
+        locked_spec.dependencies.reject! { |dep| dep.name == spec.name && dep.requirement.satisfied_by?(spec.version) }
+        deleted_deps&.each do |dep|
+          dep_to_requirements[locked_spec.name] = dep.requirement.to_s
+        end
+      end
```

Comment:
Pass one, we loop on all of these locked_specs, and we can their dependencies to try to see if they meet our current spec or not. If so, delete it (and store the requirements in a hash for reporting purposes).

For reference, spec (from embedded_gems_to_activate) looks like

```ruby
Gem::Specification.new do |s|
      s.name = "json"
      s.version = Gem::Version.new("2.7.2")
...
```

**locked_spec** looks like this

```ruby
locked_spec
 => 
#<Bundler::LazySpecification:0x00007f52209429f0                  
 @dependencies=                                                  
  [Gem::Dependency.new("json", Gem::Requirement.new(["~> 2.3"]), :runtime),
   Gem::Dependency.new("parallel", Gem::Requirement.new(["~> 1.10"]), :runtime),
   Gem::Dependency.new("parser", Gem::Requirement.new([">= 3.2.0.0"]), :runtime),
   Gem::Dependency.new("rainbow", Gem::Requirement.new([">= 2.2.2", "< 4.0"]), :runtime),
   Gem::Dependency.new("regexp_parser", Gem::Requirement.new([">= 1.8", "< 3.0"]), :runtime),
   Gem::Dependency.new("rexml", Gem::Requirement.new([">= 3.2.5", "< 4.0"]), :runtime),
   Gem::Dependency.new("rubocop-ast", Gem::Requirement.new([">= 1.28.0", "< 2.0"]), :runtime),
   Gem::Dependency.new("ruby-progressbar", Gem::Requirement.new(["~> 1.7"]), :runtime),
   Gem::Dependency.new("unicode-display_width", Gem::Requirement.new([">= 2.4.0", "< 3.0"]), :runtime)],
 @force_ruby_platform=false,
 @full_name="rubocop-1.52.0",
 @name="rubocop",
 @platform="ruby",
 @source=#<Bundler::Source::Rubygems:0x30140 locally installed gems>,
 @use_exact_resolved_specifications=true,
 @version=Gem::Version.new("1.52.0")>
```